packages/vm/src/imports.rs

```diff
@@ -150,19 +145,32 @@ macro_rules! write_region {
 }
 
 /// Reads a storage entry from the VM's storage into Wasm memory
-pub fn do_read<S: Storage, Q: Querier>(
-    ctx: &mut Ctx,
-    key_ptr: u32,
-    value_ptr: u32,
-) -> VmResult<i32> {
-    let key = read_region!(ctx, key_ptr, MAX_LENGTH_DB_KEY);
+pub fn do_read<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32) -> VmResult<u32> {
+    let key = read_region(ctx, key_ptr, MAX_LENGTH_DB_KEY)?;
     // `Ok(expr?)` used to convert the error variant.
     let value: Option<Vec<u8>> =
         with_storage_from_context::<S, Q, _, _>(ctx, |store| Ok(store.get(&key)?))?;
-    Ok(match value {
-        Some(buf) => write_region!(ctx, value_ptr, &buf),
-        None => errors::read::KEY_DOES_NOT_EXIST,
-    })
+
+    let out_data = match value {
+        Some(data) => data,
+        None => return Ok(0),
+    };
+
+    let out_ptr = with_func_from_context::<S, Q, u32, u32, _, _>(ctx, "allocate", |allocate| {
+        let out_size = to_u32(out_data.len())?;
+        let ptr = allocate.call(out_size)?;
+        Ok(ptr)
+    })?;
+
+    // TODO: is this a legal pointer in Wasm or can we blame the contract?
+    if out_ptr == 0 {
+        return Err(make_generic_err(
+            "Got a null Wasm pointer from allocate(). We cannot use it here as it represents _key does not exist_ in this context.",
+        ));
+    }
```

Comment:
I believe that it should not be returned from a correct implementation of `allocate`, so this check is correct.

Reply:
OKay, it is not so easy. What I found out at https://stackoverflow.com/q/62042733/2013738 is
- 0 is a valid memory location
- `std::ptr::null::<T>()` returns the value `0` when compiled to Wasm
- Vec guarantees to not allocate at address null: _The pointer will never be null, so this type is null-pointer-optimized._ However, this does not necessarily apply for all heap allocations

I think it is okay if we require non-0 return values from the allocate export and the contract has to ensure to not allocate at position 0.

Reply:
then yeah, in this case it's safe :+1: 
I think all heap pointers in rust are implemented with this property in practice. :)